### ISSUE TITLE
How to connect ones app to the deployed SMART on FHIR sample?

### ISSUE THREAD

**Issue by Skopfy** [Azure-Samples/azure-health-data-and-ai-samples#171]
Heyho!

Just managed to deploy the SMART on FHIR sample (thanks to the PR which explained the additional steps with ones own FHIR server).
Now i am trying to connect the sample to my application (standalone launch). Ofc, I get it that you cannot help with my custom environment (angular).
 Is there a sample app which shows one way how to do this in code? I do not even have a clue how to construct the auth request after fetching the endpoint via the discovery endpoint (that's the only step I got). I read through the implementation guide and tried some GET request to auth. Without params, its giving me a 400. With params, i am getting 500. The error messages in Azure look to obscure for me to find where I fail.

How do the individual requests look like for each step?
Is it possible to reproduce/test all the http requests with postman?

As a a still somewhat beginner in security/webdev, Id appreciate the help or some tips how to continue! :)

Best Regards
Skopfy

Edit: here is one example error I get after pressing continue here: Its giving me the error 
`Message": "Required launch parameters missing. {\"ResponseType\":\"code\",\"ClientId\":\"my-client-id\",\"RedirectUri\":null,\"Scope\":\"\",\"State\":null,\"Audience\":\"https://my-app.onmicrosoft.com/\",\"CodeChallenge\":null,\"CodeChallengeMethod\":null,\"Launch\":null,\"Prompt\":\"consent\",\"LoginHint\":\"some-random-string\"}",`
i requested the scopes "launch/patient patient/*.rs openid fhirUser"
![Screenshot from 2024-02-14 17-46-08](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/f2da7996-4df6-457d-9984-16ca74b828ad)

**Comment by evachen96** [comment]
Hi @Skopfy, thanks for your patience! You can use the attached sample Postman environment and collection as an example to construct the authorization request:

1. Import the collection and environment file in Postman.
2. Fill the required value for environment variables such as APIM url, tenant id, client id etc. (The values to be filled are specified in the environment template).
3. To generate a token, navigate to the Authorization tab in the 'fhir-proxy-smart-client-calls-sample-v2' collection.
4. Use this token to access FHIR resources based on the scopes specified in the environment variable.
5. Queries for patient and observation resources are currently added.
 
[fhir-proxy-smart-client-calls-sample-v2.postman_collection (1).json](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/files/14466365/fhir-proxy-smart-client-calls-sample-v2.postman_collection.1.json)
[apim-smart-env-sample.postman_environment (1).json](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/files/14466366/apim-smart-env-sample.postman_environment.1.json)

Let us know if this helps!

**Comment by Skopfy** [comment]
Thanks @evachen96 for the tips! :)
I pinpointed the error towards the scopes. They are correctly put under "Requested access", but not under "Selected access". I put some log statements into the files and deployed them. Apparently some requests towards MS Graph are madebut they give back an empty scope Array. In this file
https://github.dev/Azure-Samples/azure-health-data-and-ai-samples/blob/main/samples/smartonfhir/src/SMARTCustomOperations.AzureAuth/Filters/AuthorizeInputFilter.cs
Probably I did something wrong during deployment.
You have any idea what my error could be?

**Comment by evachen96** [comment]
@Skopfy  I'll have the team take a look!

**Comment by evachen96** [comment]
Hi @Skopfy, we'll need some additional information to understand more: 

- What scopes are you requesting through Postman request? (Screenshot may help)
- List of API permissions added in client application (e.g. Standalone app)
- Youcan test by requesting below scopes once:
launch/patient openid fhirUser offline_access patient/Medication.read 
patient/AllergyIntolerance.read patient/CarePlan.read patient/CareTeam.read 
patient/Condition.read patient/Device.read patient/DiagnosticReport.read 
patient/DocumentReference.read patient/Encounter.read patient/Goal.read 
patient/Immunization.read patient/Location.read patient/MedicationRequest.read 
patient/Observation.read patient/Organization.read patient/Patient.read 
patient/Practitioner.read patient/Procedure.read patient/Provenance.read 
patient/PractitionerRole.read

- If the issue persists, youcan share error logs with us using below steps to learn more about the issue:

1. Navigate to resource group created for sample.
2. Navigate to Application Insights
3. In left pane click on ‘Transaction Search’.
4. Set time duration for logs click on refresh button.
5. All the traces/exception will be listed.

Thanks!

**Comment by Skopfy** [comment]
Hey @evachen96 

of course :)
![Screenshot from 2024-03-12 09-46-53](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/6163359a-f7c5-408a-9300-ef669f4f967b)

Here the API Permissions of my SMART Auth Frontend
![Screenshot from 2024-03-12 09-51-41](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/27f61017-d330-48c7-ae2e-6f079bd396ab)

And here for the App
![image](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/71416d2d-f3b0-4291-8e26-09ddb6bd74f3)

I tried out your scopes, but same error
First I get these requested scopes, but selected ones are not showing up
![image](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/7539a23d-43e3-4a29-8705-95f1730ba6ca)

Heres the App Insights error
![image](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/80f96771-bc74-4299-96c7-164e476d4334)

**Comment by evachen96** [comment]
Hi @Skopfy -

The team has reviewed the screenshots, our primary findings are:
1.  The request URL for the authorize endpoint is not properly formatted with all required parameters. Are you following the Postman collection and environment variables and steps that we provided previously? https://github.com/Azure-Samples/azure-health-data-and-ai-samples/issues/171#issuecomment-1973977708 
2. The Scopes under Select Access are not visible because the required permissions for FHIR resource app has not been added to the API permission section of the End application as below (in this case, an Angular app).
![image](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/33711652/bf5560fb-1230-49a3-9b05-52b55c7a0e2c)

You can add permissions in the app by referring how standalone app are configured here: https://github.com/Azure-Samples/azure-health-data-and-ai-samples/blob/main/samples/patientandpopulationservices-smartonfhir-oncg10/docs/ad-apps/inferno-test-app-registration.md#:~:text=Patient%20Standalone%20Confidential%20Client%20/%20Public%20Client%20Applications.

**Comment by Skopfy** [comment]
Hey @evachen96,

thanks to your tips i progressed somewhat. The error was indeed that the API permissions had to be updated. Now they are like this:
![image](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/a62c1a77-f27f-451a-9614-b7e456d156e1)

I think it would be helpful for others if you put this in the documentation for the "smartonfhir" sample, as the link your provided belongs to another sample. For someone not so proficient with EntraID/SoF this is not so obvious to find out on oneself.
__

I solved another error, a redirect uri mismatch (request was localhost:4200 & in smart app the apim auth/context/ address)
If i put the apim address in my request it works.
I also tried your postman request with callback_url https://oauth.pstmn.io/v1/callback (preconfigured)
Why is this needed? How do I get the redirect to my app (localhost:4200 for testing)?

The next error I get is the following, happening after i receive the auth code with params code, state, session_state (all three encoded in URL)

![image](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/a7129e01-dd51-44d3-8f1e-84461ba02cd0)

The code (AppContext.tsx) has the message in this line:
```
if (applicationId == undefined || requestedScopes == undefined) {
    if (error == undefined) {
      displayError("Missing required parameters in the URL.", "client_id and scope are required.");
    }
  }

```
Postman request is the following
![Screenshot from 2024-03-21 14-05-47](https://github.com/Azure-Samples/azure-health-data-and-ai-samples/assets/11667515/30d620a9-f25a-4d54-a17f-1eb28889e4c5)

Do you have any idea what could be missing here? 

I also tried it with your postman environment, logging in as a test user. I make it until the token is to be acquired on apim.net/auth/token, where a secret is to be provided. I left it empty so an error is thrown because the client secret is missing. I do not understand why it is needed here, I want to test the auth code flow for a test user (patient/practitioner) and not client credentials flow for an app. Maybe I am missing knowledge on the OAuth flow here though ?

**Comment by evachen96** [comment]
Hi @Skopfy! We've added the Postman collection to the repo. This is a sample that you can modify and use as needed. To use the sample with postman, you can follow the steps mentioned at https://github.com/Azure-Samples/azure-health-data-and-ai-samples/blob/main/samples/smartonfhir/docs/postman/configure-postman.md.

Below are answers to your questions:

1.  How do I get the redirect to my app (localhost:4200 for testing)?
- You can try by adding redirect URL (web platform) as localhost:4200 in standalone application registration (app registered for angular app in this case).
 
2.  Which client id and client secret to be added in postman collection?
- Client Id and client secret of newly created standalone application registration (app registered for angular app in this case). You can create this app and add API permissions by following https://github.com/Azure-Samples/azure-health-data-and-ai-samples/blob/main/samples/patientandpopulationservices-smartonfhir-oncg10/docs/ad-apps/inferno-test-app-registration.md#:~:text=Patient%20Standalone%20Confidential%20Client%20/%20Public%20Client%20Applications.
 
3. Do you also use PKCE?
- No, we don’t use PKCE. We also don’t have code_challenge and code_challenge_method parameter in our postman collection.
 
4. In OAuth flow, why client secret is needed even though we are using test user’s account for sign in?
-  When you authenticate a user, you're essentially asking to grant permissions to your application to access certain resources on their behalf. This is done through OAuth 2.0 flow.
-  Even though you're using a user account for sign-in (for example, through Postman), the client ID and client secret of your registered application are still needed. They are used to authenticate your application with Azure AD during the OAuth 2.0 flow. This allows Azure AD to verify that the requests coming from your application are legitimate and that it has the necessary permissions to access the requested resources. 

Hope this helps!

**Comment by Skopfy** [comment]
Hey @evachen96!
 Thanks to your answer I finally got one of my mistakes. I did not register an extra App Registration for an Angular app additionally to  the SMART Auth FE and the fhir resource app. 

I read through the following tutorial to integrate the deployment with my Angular app. Therefore I am reading the following MS tutorial:
https://learn.microsoft.com/en-us/entra/identity-platform/tutorial-v2-angular-auth-code

The default redirect to MS Entra ID was relatively easy, but I don't know how to modify the parameters/code to integrate it with this AHDS sample. For example, how to change the msal config such that the smart auth code / token endpoints are used? 
Is there any repo/code which already did this?
